### ISSUE TITLE
Jeffreys Prior

### ISSUE THREAD

**Issue by brethvoice** [optuna/optuna#1972]
Implement Jeffreys prior as option for continuous uniformly distributed hyper-parameters.

## Motivation

I implemented this and it seemed to make Optuna studies work faster and more smartly through the continuous search spaces.

## Description

As an optional argument for `trial.suggest_uniform()` method, add a flag which would have the same effect as the following example code from the objective function:

```
# Instantiate a random number generator
rg = numpy.random.default_rng()
# Implement Jeffreys prior for continuous uniform hyper-parameter
hyper_parameter = rg.beta(0.5, 0.5) * trial.suggest_uniform(
    'hyper_parameter',
    0,
    1,
)
```

## Additional context

Not everyone likes Jeffreys priors, so it would be best to make it completely optional.  See the following repository for example of implementation:
https://github.com/brethvoice/optuna_demo_MNIST

**Comment by HideakiImamura** [comment]
Thanks for the interesting feature proposal. At the first glance, I think this feature can be implemented with the option of `optuna.samplers.RandomSampler`. @brethvoice What do you think?

**Comment by brethvoice** [comment]
> Thanks for the interesting feature proposal. At the first glance, I think this feature can be implemented with the option of `optuna.samplers.RandomSampler`. @brethvoice What do you think?

@HideakiImamura I do not think it should be an aspect of the sampler, because I plan to use it with `optuna.samplers.RandomSampler`, for example.  I would rather give people the option with a new `suggest_uniform_jeffreys_prior` method of a trial, because it only applies to the case where the hyper-parameter is continuous.

Unfortunately, I am not sure where to implement this in the code, so I am unable to do a pull request at this time.  The use of `numpy.random.default_rng` is only in my suggestion for visual effect; it would need to be integrated with whatever code you are using for random sample generation which I assumed would be similar.

**Comment by brethvoice** [comment]
Unfortunately I am not actually sure multiplying by the Beta(0.5, 0.5) factor for any continuous parameter will correctly implement a Jeffrey's prior.  I suggest running an experiment to see whether an Optuna study converges more quickly in all of the following cases:

1.  Parameter is always positive, left side of interval is 0, and right side of interval is bounded.
2.  Parameter can be either positive or negative with left and right sides bounded.
3.  Parameter is always negative, left side of interval is bounded, and right side of interval is 0.
4.  Parameter is between 0 and 1 (this is the case where I think you will certainly have success).

**Comment by brethvoice** [comment]
I am currently testing case 1 (which includes case 4) above over a multi-day set of Optuna studies.  I will give you results in less than 48 hours as long as there are no interruptions to the studies (I had memory issues crop up last weekend, probably related to an attempt to set `multiprocessing=True` within the `model.fit` method of Tensorflow's Keras.

**Comment by norihitoishida** [comment]
Hi @brethvoice 
Thank you for introducing Jeffreys Prior. It was very educational for me. 
Could you please tell me about the progress of the experiment? I'm interested in it!

**Comment by brethvoice** [comment]
> Hi @brethvoice
> Thank you for introducing Jeffreys Prior. It is very educational for me.
> Could you please tell me about the progress of the experiment? I'm interested in it!

@norihitoishida I finished a preliminary experiment with inconclusive results.  I am thinking I will kick off a full experiment soon and let you know so you can reproduce it on your end.

**Comment by brethvoice** [comment]
@norihitoishida I have the results of an experiment covering cases 1 and 4 above.  Example hyper-parameters covering case 4 were easy to find because that is the most common type of continuous hyper-parameter; case 1 was harder to find but I used the initial learning rate for an Adam optimizer in `tensorflow.keras` exponential [learning](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/ExponentialDecay) rate decay/schedule.

The results are encouraging.  I was able to get a slightly higher classification accuracy on the MNIST test set using Jeffreys priors (96.38%) vs. not using them (96.19%).

NOTE:  I did use a median pruner after the first 50 trials (500 trials completed).  I encourage you to replicate these results yourself if possible, and then incorporate Jeffreys prior as an option when specifying the sampler (perhaps?) that would be experimental at first and restricted only to cases 1 and 4 (non-negative continuous hyper-parameter).

@HideakiImamura I was wondering if it might more sense to incorporate the experimental option as part of the `trial.suggest_uniform()` call instead of during instantiation of the sampler.  This would allow for greater levels of experimentation in support of continuing research.

**Comment by brethvoice** [comment]
> @norihitoishida I have the results of an experiment covering cases 1 and 4 above. The code I used is in my Github [repo](https://github.com/brethvoice/optuna_demo_MNIST).
> 
> Example hyper-parameters covering case 4 were easy to find because that is the most common type of continuous hyper-parameter; case 1 was harder to find but I used the initial learning rate for an Adam optimizer in `tensorflow.keras` exponential [learning](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/ExponentialDecay) rate decay/schedule.
> 
> The results are encouraging. I was able to get a slightly higher classification accuracy on the MNIST test set using Jeffreys priors (96.38%) vs. not using them (96.19%).
> 
> NOTE: I did use a median pruner after the first 50 trials (500 trials completed). I encourage you to replicate these results yourself if possible, and then incorporate Jeffreys prior as an option when specifying the sampler (perhaps?) that would be experimental at first and restricted only to cases 1 and 4 (non-negative continuous hyper-parameter).
> 
> @HideakiImamura I was wondering if it might more sense to incorporate the experimental option as part of the `trial.suggest_uniform()` call instead of during instantiation of the sampler. This would allow for greater levels of experimentation in support of continuing research.

UPDATE:  I am re-starting the experiment!  I removed the learning rate schedule and amsgrad options for the Adam optimizer.

**Comment by brethvoice** [comment]
> UPDATE: I am re-starting the experiment! I removed the learning rate schedule and amsgrad options for the optimizers.

After some thought, my experiments thus far are not truly using Jeffreys priors in their intended sense.  The intent of a Jeffreys prior is only for use on the first iteration of an experiment before there is a posterior distribution.

According to this [article](https://medium.com/subex-ai-labs/efficient-hyperparameter-optimization-for-xgboost-model-using-optuna-3ee9a02566b1), Optuna is Bayesian; therefore only on the initial trial would a Jeffreys prior be useful.  Therefore I have modified the objective functions to reflect this, and also removed pruning to enable fully valid comparisons over a four day period.

The objective of the study is to determine whether a Jeffreys prior on the initial trial will cause faster convergence.  Therefore a comparison based on speed alone is sufficient.

**Comment by brethvoice** [comment]
@HideakiImamura @norihitoishida I just completed an experiment using AdaBelief vs. Adam as the optimizer, and no pruning.  I used a Jeffreys prior only in the first trial of two studies (one each for Adam and AdaBelief) and did not use it on the other two studies.

I failed to let Optuna suggest values for the `epsilon` parameter along with all of the other continuous ones.  But the Bayesian nature of AdaBelief appears to give an advantage in speed over Adam, and using a Jeffreys prior together with AdaBelief appears to give an advantage in test accuracy on the MNIST data set.

Based on these results I believe it is worth your time to implement an experimental option for a Jeffreys prior on the first trial of a study when suggesting a continous, non-negative hyper-parameter.  If time allows, I also plan to add in adversarial training and allow Optuna to change that parameter along with those of the optimizer.

**Comment by norihitoishida** [comment]
Thanks for sharing, @brethvoice !
> using a Jeffreys prior together with AdaBelief appears to give an advantage in test accuracy on the MNIST data set.

I want to reproduce the experiment. Could you please tell me your experimental's environment and quantitative result ?

**Comment by brethvoice** [comment]
> Thanks for sharing, @brethvoice !
> 
> > using a Jeffreys prior together with AdaBelief appears to give an advantage in test accuracy on the MNIST data set.
> 
> I want to reproduce the experiment. Could you please tell me your experimental's environment and quantitative result ?

I am using Python 3.6.9, tensorflow 2.3.1, numpy 1.18.5, optuna 2.3.0, and adabelief_tf 0.1.0.  If that is not enough information let me know!

**Comment by brethvoice** [comment]
@norihitoishida the quantitative results with AdaBelief were that test accuracy on MNIST was 96.44% with a Jeffreys prior, and 96.15% without using one.  Please note that the Jeffreys prior was only used on the first trial of the study, and only for these continuous hyper-parameters of AdaBelief:  learning rate, beta_1, and beta_2.

All other hyper-parameters were optimized by Optuna to be identical:
- number of conv 2d filters: 8 (max allowed to suggest)
- first x second conv 2d kernels:  2 x 2
- conv 2d layer activation function: ReLU
- number of hidden conv 2d layers: 1
- batch size:  1,024 (for training and also evaluation at end of objective function)

It is worth noting that AdaBelief reached its optimal setting on the 223rd trial out of 500, whereas Adam took only 106 trials out of the 500.  These experiments also did not allow epsilon to vary, but instead it was set as epsilon() accessed by the following:
`from tensorflow.keras.backend import epsilon`.

**Comment by brethvoice** [comment]
> @norihitoishida the quantitative results with AdaBelief were that test accuracy on MNIST was 96.44% with a Jeffreys prior, and 96.15% without using one. Please note that the Jeffreys prior was only used on the first trial of the study, and only for these continuous hyper-parameters of AdaBelief: learning rate, beta_1, and beta_2.
> 
> All other hyper-parameters were optimized by Optuna to be identical:
> 
> * number of conv 2d filters: 8 (max allowed to suggest)
> * first x second conv 2d kernels:  2 x 2
> * conv 2d layer activation function: ReLU
> * number of hidden conv 2d layers: 1
> * batch size:  1,024 (for training and also evaluation at end of objective function)
> 
> It is worth noting that AdaBelief reached its optimal setting on the 223rd trial out of 500, whereas Adam took only 106 trials out of the 500. These experiments also did not allow epsilon to vary, but instead it was set as epsilon() accessed by the following:
> `from tensorflow.keras.backend import epsilon`.

@norihitoishida please let me know if you decide to include the epsilon hyper-parameter in your experiment(s).  The default in tensorflow is 1e-7, which is the same as what I used according to [this](https://www.tensorflow.org/api_docs/python/tf/keras/backend/epsilon) page.

I think it is worth varying the value of epsilon is because of articles I have read where it can have a large effect.  My current plan is to let it vary between 0 and 1 using `trial.suggest_uniform()` within the objective function.

NOTE:  Implementing Jeffreys prior on the first trial only was done inside an if else structure checking whether `trial.number == 0` and if True, I multiply each return from `trial.suggest_uniform()` by `rg.beta(0.5, 0.5)` based on the following line of code at the top of the file:

```
from numpy.random import default_rng as random_generator_instantiator
```

followed by this line of code at the top of the objective function definition:
```
def objective(trial):
    rg = random_generator_instantiator()
```

**Comment by norihitoishida** [comment]
@brethvoice 
Thank you for sharing this very valuable information :+1: 
Given the resource of GPU, and the nature of Jeffreys Prior that is enabled only on the first trial, I will consider using [`kurobako` ](https://github.com/sile/kurobako) to do an experiment to see the Jeffreys Prior's effect before I reproduce your experiment. (My GPU can't be available for a while because of another experiment .)

**Comment by brethvoice** [comment]
> @brethvoice
> Thank you for sharing this very valuable information 👍
> Given the resource of GPU, and the nature of Jeffreys Prior that is enabled only on the first trial, I will consider using [`kurobako` ](https://github.com/sile/kurobako) to do an experiment to see the Jeffreys Prior's effect before I reproduce your experiment. (My GPU can't be available for a while because of another experiment .)

@norihitoishida I am currently running without GPU support as well; I have set the studies to run for 3.75 days before starting the last trial; hopefully that will be enough time to replicate!  Please note there was a slight re-configuration of PrunableEvaluateMNIST.py (and that I am also not allowing pruning during the experiment) so please re-pull my code if you are using it:
https://github.com/brethvoice/optuna_demo_MNIST

**Comment by norihitoishida** [comment]
Thanks @brethvoice 
If users use Optuna with TF, they will probably use the GPU.
I would like to report the experiment under conditions close to the real scenario, and the results quantitatively.
Best regards.

**Comment by brethvoice** [comment]
@norihitoishida I am using GPU support with Tensorflow now; the study will complete in about four days.  In the meantime, the `adabelief_tf` module was upgraded to version 0.2 so I am using that as well.

**Comment by brethvoice** [comment]
> @norihitoishida I am using GPU support with Tensorflow now; the study will complete in about four days. In the meantime, the `adabelief_tf` module was upgraded to version 0.2 so I am using that as well.

To allow the models to truly compete with each other I might need to train on the entire 60,000-image "training" data set, with hyper-parameter settings selected via Optuna, validating on the test data at the end of each epoch, and report the best trained epoch's performance as the final result of this experiment for each optimizer, both with and without using a Jeffreys prior to select the continuous hyper-parameters.

**Comment by brethvoice** [comment]
> > @norihitoishida I am using GPU support with Tensorflow now; the study will complete in about four days. In the meantime, the `adabelief_tf` module was upgraded to version 0.2 so I am using that as well.
> 
> To allow the models to truly compete with each other I might need to train on the entire 60,000-image "training" data set, with hyper-parameter settings selected via Optuna, validating on the test data at the end of each epoch, and report the best trained epoch's performance as the final result of this experiment for each optimizer, both with and without using a Jeffreys prior to select the continuous hyper-parameters.

@norihitoishida the re-do of the experiment (without training models on all available data) had all worse performance with Jeffreys priors on the first trial for continuous hyper-parameters (multiplied the output of trial.suggest_uniform by a Beta [0.5, 0.5] randomly generated number).  I am repeating one more time just to confirm,  at which point I will train the models on all available data with early stopping based on test set validation, and report on that performance using the suggested hyper-parameters from the Optuna studies.  This time I am setting `rectify=False` for AdaBelief at the recommendation of its developer.

**Comment by brethvoice** [comment]
I implemented median pruning, with warmup at 50% of max epochs per trial and startup before pruning at 50% of max trials.  I did this because I saw that 100 epochs per trial was not enough, and to prevent spending too much time on unpromising trials.

**Comment by brethvoice** [comment]
> I implemented median pruning, with warmup at 50% of max epochs per trial and startup before pruning at 50% of max trials. I did this because I saw that 100 epochs per trial was not enough, and to prevent spending too much time on unpromising trials.

@norihitoishida the results are mixed from the four Optuna studies.  I will now train the models on all available data with early stopping based on test set validation, and report on that performance using the suggested hyper-parameters from the Optuna studies, and report the final results of the competition.

Surprisingly, with AdaBelief optimizer with Jeffreys priors on the first trial performed better during the Optuna studies than without a Jeffreys prior.  The reverse was true for the Adam optimizer; I did notice, however, that convergence to optimal settings for the optimizer and batch size occurred more quickly in both cases where the Jeffreys priors were used on the first trial.

**Comment by brethvoice** [comment]
After some thought about the "n_startup_trials" parameter of the TPE sampler, I was wrong.  It also does not apply to the RandomSampler, which is not a Bayesian search method, so there would need to be an entire new sampler created.  I am going to close this issue but if anyone wants to look into this in the future with me, please let me know!
